src/ray/core_worker/reference_count.cc

```diff
@@ -101,6 +99,40 @@ bool ReferenceCounter::GetOwner(const ObjectID &object_id, TaskID *owner_id,
   }
 }
 
+void ReferenceCounter::DeleteReferences(const std::vector<ObjectID> &object_ids) {
+  absl::MutexLock lock(&mutex_);
+  for (const ObjectID &object_id : object_ids) {
+    auto entry = object_id_refs_.find(object_id);
+    if (entry == object_id_refs_.end()) {
+      return;
+    }
+    DeleteReferenceInternal(entry, nullptr);
+  }
+}
+
+void ReferenceCounter::DeleteReferenceInternal(
+    absl::flat_hash_map<ObjectID, Reference>::iterator entry,
+    std::vector<ObjectID> *deleted) {
+  if (entry->second.on_delete) {
+    entry->second.on_delete(entry->first);
+  }
+  if (deleted) {
+    deleted->push_back(entry->first);
+  }
+  object_id_refs_.erase(entry);
+}
+
+bool ReferenceCounter::SetDeleteCallback(
+    const ObjectID &object_id, const std::function<void(const ObjectID &)> callback) {
+  absl::MutexLock lock(&mutex_);
+  auto it = object_id_refs_.find(object_id);
+  if (it == object_id_refs_.end()) {
+    return false;
+  }
+  it->second.on_delete = callback;
```

Comment:
Seems like you need a list of callbacks here.

Reply:
Not currently. Might in the future but will cross that bridge when we come to it

Reply:
Added a check to be sure we don't overwrite them. This will only ever be called once per object.

Reply:
Won't you hit this if two workers wait for eviction on the same object?